Fred initial implement

pavel-georgiev opened the pull request.

linkous8 committed:
First pass implementation
- adjust passing pytests
- measure ready for testing

linkous8 committed:
Update README.md

dependencies to list

linkous8 committed:
Updated config 'value' => 'default'
- measure pytests passing

linkous8 committed:
Implemented expected_duration, progress messages

linkous8 committed:
Set value to default if not in state

linkous8 committed:
Remove default from describe

linkous8 committed:
Populate default state from config when needed

linkous8 committed:
Dockerfile done, tested
- Yaml load => safe_load

linkous8 committed:
Updated config with command for docker testing

linkous8 committed:
Added setting type based validation to adjust

pavel-georgiev commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+    https://raw.githubusercontent.com/opsani/servo/master/adjust.py \
+    https://raw.githubusercontent.com/opsani/servo/master/measure.py \
+    https://raw.githubusercontent.com/opsani/servo/master/state_store.py \
+    https://raw.githubusercontent.com/opsani/servo-batch/fred-initial-implement/adjust \
```
Please change these to be ./adjust and ./measure, no need for external ref.

pavel-georgiev commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+    /servo/
+
+RUN chmod a+rwx /servo/adjust /servo/measure /servo/servo
+RUN chmod a+rw /servo/measure.py /servo/adjust.py /servo/state_store.py
```
These don't need to be executable

linkous8 replied:
Was able to reduce the above to the following:

`RUN chmod a+x /servo/adjust /servo/measure`

If adjust and measure are not made executable, the following error occurs:

```
DRIVER REQUEST: ./adjust None
Traceback (most recent call last):
  File "servo", line 361, in exec_command
    v = dict(descriptor=describe(), status='ok')
  File "servo", line 117, in describe
    envd = run_driver(DFLT_ADJUST_DRIVER, args.app_id, describe='--query')
  File "servo", line 194, in run_driver
    proc = subprocess.Popen(cmd, bufsize=0, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)
  File "/usr/local/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/usr/local/lib/python3.6/subprocess.py", line 1364, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: './adjust'
DESCRIBE {}: will return {'status': 'failed', 'message': "PermissionError: [Errno 13] Permission denied: './adjust'"}. 
```

pavel-georgiev replied:
My commend was on the following line:

RUN chmod a+rw /servo/measure.py /servo/adjust.py /servo/state_store.py

Those should not need to be executable

pavel-georgiev commented on README.md:
```diff
@@ -20,6 +20,10 @@ See sample `config.yaml` below for examples.
 
 TBD: Dependencies (i.e. adjust.py, measure.py, etc.)
```
Please change that to be an actual section with external dependencies.

pavel-georgiev commented on adjust:
```diff
@@ -0,0 +1,163 @@
+import sys
+
+# if this is running in the pytest suite, symlink destination names have been adjusted to make importing easier
+if 'pytest' in sys.modules:
```
Can you change this so that the test has all the logic on how to deal with imports rather than the actual driver.

pavel-georgiev commented on adjust:
```diff
@@ -0,0 +1,163 @@
+        cfg = d[DRIVER_NAME]
+        assert cfg and isinstance(cfg, dict), "Config located at {} was malformed; must provide non-empty `{}` dictionary, found: {}".format(CONFIG_PATH, DRIVER_NAME, cfg)
+
+        bad_top_keys = cfg.keys() - {'application', 'command', 'expected_duration', 'metrics'} # TODO: warmup, delay?
```
No need to warmup and delay

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+        assert isinstance(command, str), 'The `command` attribute in `batch` section must be a string, found {}'.format(type(command).__name__)
+
+        ed = config.get('expected_duration')
+        if ed:
```
This still allows me to set expected_duration=[]

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+    def _print_progress_wrapper(self):
+        if self.start_time:
+            elapsed_time = time.time() - self.start_time
+            self.progress = int(round(elapsed_time / self.config.get('expected_duration', 3600) * 100))
```
Please move the 3600 to a constant

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+
+        return metrics
+
+    def _print_progress_wrapper(self):
```
Does this need to be a wrapper? Can it simply overwrite the parent's method

linkous8 replied:
Given there is no need modify the parent functionality beyond setting the progress percentage, I found it simplest to call the parent's method from the wrapper here

https://github.com/opsani/servo-batch/blob/111f6dca937517313110caa7bc65a0dd13825dfa/measure#L113

And then override the print_progress_message of the driver class here

https://github.com/opsani/servo-batch/blob/111f6dca937517313110caa7bc65a0dd13825dfa/measure#L34

If I were to overwrite the entire function, I would need to redundantly implement all the code contained within this method

https://github.com/opsani/servo/blob/0947d14f2b23677b74960b3e76df80ba2555353c/measure.py#L151

pavel-georgiev replied:
What I had in mind was:
```
def print_progress():
       if self.start_time:
            elapsed_time = time.time() - self.start_time
            self.progress = int(round(elapsed_time / self.config.get('expected_duration', 3600) * 100))
      super().print_progress()
```
Would that not work?

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+        # Construct a state from config defaults if batch adjust has not occurred yet 
+        if not state:
+            def_state_comps = {}
+            for c_name, c_data in self.config["application"]["components"].items():
```
The "application" key can contain other things besides "components", currently just "annotations".

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+            state = { "application": { "components": def_state_comps } }
+
+        self.progress_message = "Formatting command"
+        self.print_progress()
```
This will be a very quick operation, probably not worth sending progress on it. "running command" should be sufficient.

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+        self.print_progress()
+        
+        try:
+            formatted = self.config['command'].format(state=state, control=control)
```
Please include an example in the README on how to use literal {} characters in the command (i.e. without substitution).

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+        self.print_progress()
+
+        ret_metrics = {}
+        if 'total_runtime' in in_metrics:
```
total_runtime should be returned always, regardless of input.

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+            in_metrics.remove('total_runtime')
+
+        # regex parse on output
+        if in_metrics:
```
I think the driver should return metrics based on config file, not what was passed as part of the input. @pnickolov what do you think?

pavel-georgiev commented on test/dummy_job.sh:
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
```
Great job including tests, please add a note in the README on how to run them.

pavel-georgiev commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+# Install dependencies
+RUN pip3 install requests PyYAML python-dateutil boto3
+
+# Change /usr/bin/python3 => /usr/local/bin/python3, since the first is older,
```
Is boto3 needed? This /usr/bin/python3 => /usr/local/bin/python3 should not be needed either.

pavel-georgiev commented on adjust:
```diff
@@ -0,0 +1,163 @@
+
+class BatchAdjustDriver(Adjust):
+
+    def _parse_cfg(self):
```
Can this be moved so that it can be reused between adjust and measure

pavel-georgiev commented on measure:
```diff
@@ -0,0 +1,181 @@
+            raise Exception('Measure input is missing "metrics" key')
+
+        control = self.input_data.get('control')
+        state = state_store.get_state()
```
Getting the state should be the same operation as `./adjust --query`, but without duplicating the code. You either reuse most the code in Adjust.query() or call `./adjust --query` (probably an overkill).

pavel-georgiev commented on adjust:
```diff
@@ -0,0 +1,163 @@
+        if state:
+            assert ("application" in state), "Invalid state: missing application key"
+
+        for c_name, c_data in cfg["application"]["components"].items():
```
"application" may include other things besides "components". A better approach would be to take the "application" key from config, replace anything that is not "components" with a matching key from the statestore (if present), then update components with the value from statestore (if present).

pavel-georgiev commented on adjust:
```diff
@@ -0,0 +1,163 @@
+        for c_name, c_data in cfg["application"]["components"].items():
+            for s_name, _ in c_data["settings"].items():
+                # Update 'value' key from state if present (and contained by config)
+                if state:
```
This 'if...else' is redundant. The try: except: code below should cover both cases.

linkous8 committed:
Resolutions for the following
- https://github.com/opsani/servo-batch/pull/1#discussion_r348281701
- https://github.com/opsani/servo-batch/pull/1#discussion_r348281759
- https://github.com/opsani/servo-batch/pull/1#discussion_r348282249
- https://github.com/opsani/servo-batch/pull/1#discussion_r348282379
- https://github.com/opsani/servo-batch/pull/1#discussion_r348283759
- https://github.com/opsani/servo-batch/pull/1#discussion_r348284081
- https://github.com/opsani/servo-batch/pull/1#discussion_r348285443
- https://github.com/opsani/servo-batch/pull/1#discussion_r348286025
- https://github.com/opsani/servo-batch/pull/1#discussion_r348287143
- https://github.com/opsani/servo-batch/pull/1#discussion_r348290258

linkous8 committed:
Updated format of settings in command config attr

linkous8 committed:
Resolutions for the following
- https://github.com/opsani/servo-batch/pull/1#discussion_r348287697
- https://github.com/opsani/servo-batch/pull/1#discussion_r348288697

linkous8 committed:
Resolutions for the following
- https://github.com/opsani/servo-batch/pull/1#discussion_r348285223
- https://github.com/opsani/servo-batch/pull/1#discussion_r348289905

linkous8 committed:
Include stdout and stderr of failed command

pavel-georgiev commented on common.py:
```diff
@@ -0,0 +1,141 @@
+    for key, val in cfg["application"].items():
+        if key == "components":
+            continue
+        try:            
```
Why is this in a try?

linkous8 replied:
So that the value from config will be used in cases where state is `None` (eg. prior to the first adjustment) as well as cases where state doesn't contain the key from the config

pavel-georgiev commented on common.py:
```diff
@@ -0,0 +1,141 @@
+    q = {
+        "components": cfg["application"]["components"]
+    }
+    if "annotations" in cfg["application"]:
```
Is this not covered by line 123 already?

pavel-georgiev commented on test/test_measure.py:
```diff
@@ -0,0 +1,47 @@
+        m.setattr(sys, 'argv', ['', '1234'])
+        m.setattr(sys, 'stdin', io.StringIO(measure_json_stdin))
+        driver = BatchMeasureDriver(cli_desc=DESC, supports_cancel=HAS_CANCEL, version=VERSION)
+        driver.run()
```
Can this test be modified to validate the output of measure.

linkous8 committed:
Updated Readme to reflect syntax update. Resolved:
- https://github.com/opsani/servo-batch/pull/1#discussion_r348281870
- https://github.com/opsani/servo-batch/pull/1#discussion_r348285690

linkous8 committed:
Pytest Instructions
- https://github.com/opsani/servo-batch/pull/1#discussion_r348286725

linkous8 committed:
Readme yaml command, yaml spacing, expected_dur

linkous8 committed:
Merge branch 'master' into fred-initial-implement

linkous8 committed:
Updated dockerfile, test config

linkous8 committed:
Merge branch 'fred-initial-implement' of https://github.com/opsani/servo-batch into fred-initial-implement

linkous8 committed:
Resolved the following:
- https://github.com/opsani/servo-batch/pull/1#discussion_r348891105
- https://github.com/opsani/servo-batch/pull/1#discussion_r348892422
- https://github.com/opsani/servo-batch/pull/1#discussion_r348893509

Commit:
Update test instructions
